File: kattis/contests/uky/uk23practice3/molecules.py
n,m = map(int,input().split())
pos = []
move = set()
for i in range(n):
    a,b = map(int,input().split())
    pos.append((a,b))
    if a == b == -1:
        move.add(i)

adj = [[] for _ in range(n)]
for _ in range(m):
    a,b = map(int,input().split())
    a-=1
    b-=1
    adj[a].append(b)
    adj[b].append(a)

miter = 10**4
for _ in range(miter):
    npos = []
    for i in range(n):
        if i not in move:
            npos.append(pos[i])
        else:
            nx=ny=0
            for nn in adj[i]:
                nx += pos[nn][0]
                ny += pos[nn][1]
            nx /= len(adj[i])
            ny /= len(adj[i])
            if pos[i][0] == pos[i][1] == -1:
                npos.append((nx,ny))
            else:
                nx = (nx+pos[i][0])/2
                ny=(ny+pos[i][1])/2
                npos.append((nx,ny))
    pos = npos
for p in pos:
    print(*p)

File: kattis/contests/uky/uk23practice3/molecules.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n,m; cin >>n>>m;
    vector<pair<double,double>> pos;
    unordered_set<int> move;
    for(int i = 0; i < n; i++) {
        int a,b;cin>>a>>b;
        pos.emplace_back(a,b);
        if (a==b &&b==-1) move.insert(i);
    }

    vector<vector<int>> adj(n);
    for(int i=0;i<m;i++) {
        int a,b;cin>>a>>b;
        a--,b--;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    int miter = 100000;
    for(int x=0;x<miter;x++){
//        cout <<x<<endl;
        vector<pair<double,double>> npos;
        for(int i = 0; i < n; i++) {
            if (move.find(i) == move.end()) npos.push_back(pos[i]);
            else {
                double nx=0,ny=0;
                for(auto nn:adj[i]) {
                    nx += pos[nn].first;
                    ny+=pos[nn].second;
                }
                nx/=1.0*adj[i].size();
                ny/=1.0*adj[i].size();
                if (pos[i].first == -1)
                    npos.emplace_back(nx,ny);
                else {
                    nx=(nx+pos[i].first)/2.0;
                    ny=(ny+pos[i].second)/2.0;
                    npos.emplace_back(nx,ny);
                }
            }
        }
        swap(pos,npos);
    }

    for(auto p : pos)
        cout << p.first << " " << p.second << endl;
}